service/consul/instancer.go

```diff
@@ -136,29 +145,32 @@ func (i *instancer) getInstances(lastIndex uint64, stop <-chan struct{}) ([]stri
 	result := make(chan response, 1)
 
 	go func() {
-		var queryOptions api.QueryOptions = i.queryOptions
-		queryOptions.WaitIndex = lastIndex
-		entries, meta, err := i.client.Service(i.service, i.tag, i.passingOnly, &queryOptions)
-		if err != nil {
-			result <- response{err: err}
-			return
-		}
+		var (
+			queryOptions api.QueryOptions = i.queryOptions
+			entries      []*api.ServiceEntry
+			meta         *api.QueryMeta
+			resp         response
+		)
 
-		if len(i.filterTags) > 0 {
-			entries = filterEntries(entries, i.filterTags)
-		}
-
-		// see: https://www.consul.io/api-docs/features/blocking#implementation-details
-		if meta == nil || meta.LastIndex < lastIndex {
-			lastIndex = 0
-		} else {
-			lastIndex = meta.LastIndex
+		queryOptions.WaitIndex = lastIndex
+		entries, meta, resp.err = i.client.Service(i.service, i.tag, i.passingOnly, &queryOptions)
+		if resp.err == nil {
+			// see: https://www.consul.io/api-docs/features/blocking#implementation-details
+			if meta == nil || meta.LastIndex < lastIndex {
+				resp.err = errIndexUnderflow
```

Comment:
Would there be any value to having metrics showing when there are boundary conditions like this occuring?

Reply:
Possibly.  Remember, though, that this is normal consul behavior.  In particular, this happens in the face of network partitioning, which is simply a temporary condition that service discovery infrastructure should be tolerant of.

If we want metrics around this, let's do it in a separate PR.